### DISCUSSION TITLE
Feature Request: Use a protocol handler for the Follow button

### DISCUSSION THREAD

**Discussion by kaleb70** [mastodon/mastodon#21320] [General]
> **Note**
> This was already being discussed at https://github.com/mastodon/mastodon/issues/19679, and I had missed that discussion when I posted this.

### Basic Idea

If you are viewing a profile on a server where you are not logged in, the Follow button would be a link that looks like this:

`web+mstdn:@kalebaxon@deacon.social?follow=1`

Unfortunately, setting up the protocol handler is browser- and platform-specific.

* Android and iOS: Mastodon client apps must be configured as the protocol handler.
* Firefox, Opera, desktop Chrome/Chromium/Edge: Use `registerProtocolHandler` and show a friendly message to the user guiding them to accept the request.
* Desktop Safari: Someone else would need to chime in on how this works.

### The Follow Button

The Follow button would execute a different script depending on whether the user is logged in.

```mermaid
graph TD
    Start([Render Follow Button]) --> LoggedIn
    LoggedIn{Logged In?}
    LoggedIn -->|Yes| DirectFollow
    LoggedIn -->|No| IndirectFollow
    DirectFollow[Script for Direct Follow]
    DirectFollow --> Done
    IndirectFollow[Script for Indirect Follow]
    IndirectFollow --> Done
    Done([Done])
```

If the user is logged in, the Follow button's behavior does not change. The change comes when the user is _not_ logged in, in which case `window.location` is set to a URL such as `web+mstdn:@kalebaxon@deacon.social?follow=1`, and an alternate UI is shown for when the protocol handler is not configured.

```mermaid
graph TD
    Start([Indirect Follow]) --> GreyOut
    GreyOut[Grey Out Follow Button] --> Link
    Link[set window.location] --> Delay
    Delay[Delay to Let New Page Load] --> ShowUI
    ShowUI[Show Alternate UI] --> Done
    Done([Done])
```

### Alternate Follow UI

The alternate Follow UI would prompt the user to enter their own Mastodon server, then redirect the user there. For example, if a user is trying to follow `@kalebaxon@deacon.social` and they enter their Mastodon server as `mastodon.social`, they would be directed to `https://mastodon.social/@kalebaxon@deacon.social?follow=1&protocolhelp=1`. This would instruct the user's own server to handle the follow request and help the user authorize the protocol handler.

**Comment by nightpool** [comment]
See conversation in #19679

**Reply by kaleb70** [reply]
Thanks. I thought I searched for it, but obviously I goofed it up somehow. 😅
